p2p/net/transport/tcp.go

```diff
@@ -0,0 +1,233 @@
+package transport
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	ma "github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/jbenet/go-multiaddr"
+	manet "github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/jbenet/go-multiaddr-net"
+	reuseport "github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/jbenet/go-reuseport"
+	context "github.com/ipfs/go-ipfs/Godeps/_workspace/src/golang.org/x/net/context"
+	lgbl "github.com/ipfs/go-ipfs/util/eventlog/loggables"
+)
+
+type TcpTransport struct {
+	dlock   sync.Mutex
+	dialers map[string]Dialer
+
+	llock     sync.Mutex
+	listeners map[string]Listener
+}
+
+func NewTCPTransport() *TcpTransport {
+	return &TcpTransport{
+		dialers:   make(map[string]Dialer),
+		listeners: make(map[string]Listener),
+	}
+}
+
+func (t *TcpTransport) Dialer(laddr ma.Multiaddr, opts ...DialOpt) (Dialer, error) {
+	t.dlock.Lock()
+	defer t.dlock.Unlock()
+	s := laddr.String()
+	d, found := t.dialers[s]
+	if found {
+		return d, nil
+	}
+	var base manet.Dialer
+
+	for _, o := range opts {
+		switch o := o.(type) {
+		case TimeoutOpt:
+			base.Timeout = o.(time.Duration)
+		default:
+			return nil, fmt.Errorf("unrecognized option: %#v", o)
+		}
+	}
+
+	tcpd, err := t.newTcpDialer(base, laddr)
+	if err != nil {
+		return nil, err
+	}
+
+	t.dialers[s] = tcpd
+	return tcpd, nil
+}
+
+func (t *TcpTransport) Listener(laddr ma.Multiaddr) (Listener, error) {
+	t.llock.Lock()
+	defer t.llock.Unlock()
+	s := laddr.String()
+	l, found := t.listeners[s]
+	if found {
+		return l, nil
+	}
+
+	list, err := manetListen(laddr)
+	if err != nil {
+		return nil, err
+	}
+
+	tlist := &tcpListener{
+		list:      list,
+		transport: t,
+	}
+
+	t.listeners[s] = tlist
+	return tlist, nil
+}
+
+func manetListen(addr ma.Multiaddr) (manet.Listener, error) {
+	network, naddr, err := manet.DialArgs(addr)
+	if err != nil {
+		return nil, err
+	}
+
+	if ReuseportIsAvailable() {
+		nl, err := reuseport.Listen(network, naddr)
```

Comment:
i think should make reuseport an option, and shouldn't use it if it isn't set. (i.e. `TCPTransport` by default shouldnt use reuseport as it may cause unexpected behavior?)